main.tf

```diff
@@ -1 +1,346 @@
+###
+### LOCALS
+###
 
+locals {
+  name_id         = "${var.name}-${random_string.this.result}"
+  bucket          = "vault-salt-formula"
+  ami_name_filter = "spel-minimal-centos-7-hvm-*.x86_64-gp2"
+  ami_name_regex  = "spel-minimal-centos-7-hvm-\\d{4}\\.\\d{2}\\.\\d{1}\\.x86_64-gp2"
+  vpc_id          = "${data.aws_subnet.lb.0.vpc_id}"
+  policy_name     = "INSTANCE_VAULT_${data.aws_caller_identity.current.account_id}"
+  ssm_root_path   = "vault/${var.environment}/${data.aws_caller_identity.current.account_id}/${var.name}"
+  plain_domain    = "${replace(element(split(":", var.vault_dns_address), 1), "////", "")}"
+  public_ip       = "${chomp(data.http.ip.body)}/32"
+  allow_inbound   = "${compact(distinct(concat(list(local.public_ip), var.additional_ips_allow_inbound)))}"
+
+  tags = {
+    Environment = "${var.environment}"
+  }
+}
+
+###
+### DATA SOURCES
+###
+
+data "aws_partition" "current" {}
+
+data "aws_caller_identity" "current" {}
+
+data "aws_region" "current" {}
+
+data "aws_ami" "this" {
+  most_recent = "true"
+
+  owners = ["${var.ami_owner}"]
+
+  name_regex = "${local.ami_name_regex}"
+
+  filter {
+    name   = "name"
+    values = ["${local.ami_name_filter}"]
+  }
+}
+
+data "http" "ip" {
+  url = "http://ipv4.icanhazip.com"
+}
+
+data "aws_subnet" "lb" {
+  count = "${length(var.lb_subnet_ids)}"
+
+  id = "${var.lb_subnet_ids[count.index]}"
+}
+
+data "aws_kms_key" "this" {
+  key_id = "${var.kms_key_id}"
+}
+
+data "template_file" "instance_policy" {
+  template = "${file("policies/iam_policy.json")}"
+
+  vars {
+    partition  = "${data.aws_partition.current.partition}"
+    region     = "${data.aws_region.current.name}"
+    account_id = "${data.aws_caller_identity.current.account_id}"
+
+    name           = "${var.name}"
+    environment    = "${var.environment}"
+    key_id         = "${data.aws_kms_key.this.key_id}"
+    dynamodb_table = "${var.dynamodb_table}"
+    bucket_name    = "${local.bucket}"
+    ssm_path       = "${local.ssm_root_path}"
+  }
+}
+
+data "aws_iam_policy_document" "instance_trust_policy" {
+  statement {
+    actions = ["sts:AssumeRole"]
+
+    principals {
+      type        = "Service"
+      identifiers = ["ec2.${var.url_suffix}"]
+    }
+  }
+}
+
+# Manage IAM Roles and Policies
```

Comment:
recommend moving iam requirements into their own module (like the s3 bucket) for better portability across environments